File: medium/2369. Check if There is a Valid Partition For The Array/tabulation.java
class Solution {
    public boolean validPartition(int[] nums) {
        int n = nums.length;
        // base case
        if (n == 2) {
            return nums[0] == nums[1];
        }
        // set initial value
        boolean[] dp = new boolean[n + 1];
        if (nums[0] == nums[1]) {
            dp[1] = true;
        }
        if ((nums[0] == nums[1] && nums[1] == nums[2]) || (nums[0] + 1 == nums[1] && nums[1] + 1 == nums[2])) {
            dp[2] = true;
        }
        // update the dp table
        for (int i = 3; i < n; i++) {
            if (nums[i] == nums[i - 1]) {
                dp[i] |= dp[i - 2];
            }
            if (nums[i] == nums[i - 1] && nums[i - 1] == nums[i - 2]) {
                dp[i] |= dp[i - 3];
            }
            if (nums[i] - 1 == nums[i - 1] && nums[i - 1] - 1 == nums[i - 2]) {
                dp[i] |= dp[i - 3];
            }
        }
        return dp[n - 1];
    }
}

File: medium/2369. Check if There is a Valid Partition For The Array/tabulation.cpp
#include <vector>

using namespace std;

class Solution
{
public:
    bool validPartition(vector<int> &nums)
    {
        vector<int> dp(nums.size() + 1, 0);
        dp[nums.size()] = 1;
        for (int i = nums.size() - 1; i >= 0; i--)
        {
            if (i + 1 < nums.size() and nums[i] == nums[i + 1])
                dp[i] = max(dp[i], dp[i + 2]);
            if (i + 2 < nums.size())
            {
                if (nums[i] == nums[i + 1] and nums[i + 1] == nums[i + 2])
                    dp[i] = max(dp[i], dp[i + 3]);
                else if (nums[i + 1] - nums[i] == 1 && nums[i + 2] - nums[i + 1] == 1)
                    dp[i] = max(dp[i], dp[i + 3]);
            }
        }

        return dp[0];
    }
};